### ISSUE TITLE
Enemy units not populating when starting a battle from MekHQ

### ISSUE THREAD

**Issue by BowtieBob** [MegaMek/mekhq#2273]
The last two scenarios that have been created have not brought any enemy units into the MegeMek battle when it starts.  My own units, and allied units are populating, but not the enemy.  These scenarios were from different missions (the last scenario of one mission, and now the first scenario of a new mission).

### Environment
What version of MekHQ does your issue apply to? 0.46.1
What operating system are you using?  Windows, Mac, or Linux? Windows
What java version are you using?  version 8 Update 271
[Nightstalkers30620327.zip](https://github.com/MegaMek/mekhq/files/5681474/Nightstalkers30620327.zip)

**Comment by BowtieBob** [comment]
After creating a new campaign in 0.47.16, I am having this same issue again.  Here is a recent campaign file, as well as the logs and custom units.
[Nightstalkers30590501.cpnx.zip](https://github.com/MegaMek/mekhq/files/5797581/Nightstalkers30590501.cpnx.zip)

**Comment by BowtieBob** [comment]
As a test, I created a completely new campaign in 0.47.16.  This time my units and the enemy reinforcements showed up, but not my ally or the main enemy force.
[My Campaign30670523.cpnx.zip](https://github.com/MegaMek/mekhq/files/5797696/My.Campaign30670523.cpnx.zip)

**Comment by NickAragua** [comment]
We're going to need log files as well.

**Comment by BowtieBob** [comment]
there are logs in the zip files

**Comment by BowtieBob** [comment]
There is a huge string of this error repeating in the logs for some reason:

ERROR [mekhq.AtBGameThread] {My Campaign} 
run(), line 112 : Thread in unknown stage

**Comment by NickAragua** [comment]
Can't reproduce this on a fresh 47.16 install or in the dev environment using any of the saves provided. My guess is that MegaMek is being a little bit slow to load, or the previous battle wasn't terminated properly.

**Comment by BowtieBob** [comment]
As another test, I went back to my old campaign and opened it in 0.47.15. It was able to generate all forces with no problem.

**Comment by sixlettervariables** [comment]
That's an interesting error where MHQ cannot connect to the MM instance it started:
```
            while (client.getLocalPlayer() == null) {
                Thread.sleep(50);
            }

            // if game is running, shouldn't do the following, so detect the
            // phase
            for (int i = 0; (i < 1000) && (client.getGame().getPhase() == IGame.Phase.PHASE_UNKNOWN); i++) {
                Thread.sleep(50);
                MekHQ.getLogger().error("Thread in unknown stage" );
            }
```

**Comment by BowtieBob** [comment]
I have this installed in a cloud storage folder so that I can access it from both my desktop and my laptop. It was working fine on 0.47.15. I tried a new install to the cloud folder of 0.47.16 with no luck. I copied the files to my desktop hard drive outside of the cloud folder and it was able to populate all the forces in the battle.

**Comment by sixlettervariables** [comment]
It sounds like IO performance is an issue.

If you're reasonably comfortable with taking a performance benchmark that would be very helpful.
1. Open MekHQ and get ready to start a game
2. Download and run JVisualVM
3. Double click on the MekHQ process
4. Go to the Sampling tab
5. Start CPU sampling 
6. In MekHQ start the game
7. Wait until it times out or loads the game
8. Return to JVisualVM and stop sampling

**Comment by BowtieBob** [comment]
[VisualVM snapshot 01-11-21.zip](https://github.com/MegaMek/mekhq/files/5799728/VisualVM.snapshot.01-11-21.zip)
here is a anapshot from visualvm

**Comment by sixlettervariables** [comment]
From the perf trace it appears there were two players and both connected. This may mean we have a race condition induced by your file system performance. Why? I don't know. We likely need to spend some time looking into how the `AtBGameThread` works and see.

Between 0.47.15 and 0.47.16 nothing regarding those were changed, and nothing regarding the server or server connection handling was changed. Why 0.47.16 suddenly has this issue and 0.46.15 does not is odd.

**Comment by BowtieBob** [comment]
[VisualVM snapshot 01-12-21.zip](https://github.com/MegaMek/mekhq/files/5805014/VisualVM.snapshot.01-12-21.zip)
Apparently I sometimes have issues with 0.47.15 as well.  I figured I would take a visualvm snapshot from 0.47.15, and while it did load the enemy forces, it did not load my allies.

**Comment by BowtieBob** [comment]
I just had this happen again while starting a battle on 0.47.17. I was running some CPU intensive analysis in the background when I started the battle and I had my task manager open. The battle included myself and three enemies. My CPU maxed out at 100% right after my units started loading. The next two enemies stayed at 0BV, but then the CPU usage dropped back down, and the last enemy loaded. 

I reloaded the battle three times, and any units that were attempting to load when the CPU was maxed wouldn't come in, but the rest would. 

This may just be an issue of needing to wait until the CPU usage is below a certain value before loading units.
